sphinx.go

```diff
@@ -660,25 +677,64 @@ func NewRouter(nodeKey *btcec.PrivateKey, net *chaincfg.Params) *Router {
 // In the case of a successful packet processing, and ProcessedPacket struct is
 // returned which houses the newly parsed packet, along with instructions on
 // what to do next.
-func (r *Router) ProcessOnionPacket(onionPkt *OnionPacket, assocData []byte) (*ProcessedPacket, error) {
-	dhKey := onionPkt.EphemeralKey
-	routeInfo := onionPkt.RoutingInfo
-	headerMac := onionPkt.HeaderMAC
+func (r *Router) ProcessOnionPacket(onionPkt *OnionPacket,
+	assocData []byte) (*ProcessedPacket, error) {
 
+	// Compute the shared secret for this onion packet.
 	sharedSecret, err := r.generateSharedSecret(onionPkt.EphemeralKey)
 	if err != nil {
 		return nil, err
 	}
 
-	// In order to mitigate replay attacks, if we've seen this particular
-	// shared secret before, cease processing and just drop this forwarding
-	// message.
-	r.RLock()
-	if _, ok := r.seenSecrets[sharedSecret]; ok {
-		r.RUnlock()
-		return nil, ErrReplayedPacket
+	// Additionally, compute the hash prefix of the shared secret, which
+	// will serve as an identifier for detecting replayed packets.
+	hashPrefix := hashSharedSecret(&sharedSecret)
+
+	// Continue to optimistically process this packet, deferring replay
+	// protection until the end to reduce the penalty of multiple IO
+	// operations.
+	packet, err := processOnionPacket(onionPkt, &sharedSecret, assocData)
+	if err != nil {
+		return nil, err
+	}
+
+	// Retrieve the packet's forwarding instructions, which will contain the
+	// CLTV at which the packet expires.
+	hopData := packet.ForwardingInstructions
+
+	// Atomically compare this hash prefix with the contents of the on-disk
+	// log, persisting it only if this entry was not detected as a replay.
+	if err := r.log.Put(&hashPrefix, hopData.OutgoingCltv); err != nil {
```

Comment:
I don't see a check (a read to the DB) to ensure that this _isn't_ a replay attempt.

Reply:
https://github.com/lightningnetwork/lightning-onion/pull/16/files/5b54e54043166771276aac54d39694a7ae784b83#diff-f0b1d467eba95bd084612c430f8984e8R344

Reply:
had forgotten that i made the operation atomic lol